### PR TITLE
Add functionality to sign the string-to-sign via Google IAM REST API

### PR DESCRIPTION
**Pull request by mruoss**

### CODE REVIEW

**Comment by mruoss** [comment]
@alexandrubagu still WIP, but feel free to leave comments.

File: lib/gcs_signed_url/crypto.ex [hunk 1]
```diff
@@ -34,14 +36,59 @@ defmodule GcsSignedUrl.Crypto do
...
+  @doc """
+  Signs the given string via the signBlob REST API
+  (see https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/signBlob)
+
+  ## Examples
+
+      iex> GcsSignedUrl.Crypto.sign_via_api("foo", %GcsSignedUrl.OAuthConfig{access_token: "..."})
+      {:ok, "1fad6186e41f577a37f56589..."}
+  """
+  @spec sign_via_api(String.t(), SignBlob.OAuthConfig.t()) :: String.t()
```
Comment (alexandrubagu) [inline lib/gcs_signed_url/crypto.ex:56]: I think a code like this is hard to follow. I would go with something like this:
```elixir
  def sign_via_api(string_to_sign, %SignBlob.OAuthConfig{} = oauth_config) do
    with response <- do_post_request(string_to_sign, oauth_config),
         {:ok, body} <- Jason.decode(response.body) do
      {:ok, body["signedBlob"]}
    end
  end

  defp do_post_request(string_to_sign, oauth_config) do
    payload = Base.encode64(string_to_sign)

    @sign_blob_http.post(
      oauth_config.service_account,
      %{payload: payload},
      Authorization: "Bearer #{oauth_config.access_token}"
    )
  end
```

As you can see I didn't pattern match on `%HTTPoison.Response` because if you want to change the library you have to change the code. 
Also you can move error mapping in other place, basically where you call this function

```elixir
with {:ok, signature} <- GcsSignedUrl.Crypto.sign_via_api(...) do
   ...
else 
   error -> format_error(error)
end

File: test/gcs_signed_url/crypto_test.exs [hunk 2]
```diff
@@ -16,4 +18,68 @@ defmodule GcsSignedUrl.CryptoTest do
...
+
+    test "returns signature upon successful response from the HTTP client", %{
+      oauth_config: oauth_config,
+      string_to_sign: string_to_sign
+    } do
+      MockSetup.sign_via_api()
+      assert {:ok, "signature"} == MUT.sign_via_api(string_to_sign, oauth_config)
+    end
+
+    test "returns error with details upon 401 response from API", %{
```
Comment (alexandrubagu) [inline test/gcs_signed_url/crypto_test.exs:41]: Feels like over testing all the status codes and all the errors. Another idea is to create an Errors module where you could have `def forbidden(message), do: {:error, ...}`, `def unauthorized(message), do: {:error, ...}`

**Comment by mruoss** [comment]
Alright, development is done. I had to refactor again quite a bit because generating this string to sign is quite complex and related. I moved the logic of creating the string to sign into the StringToSign module which anyway was not doing much before. It now returns the string to sign together with the URL it created the string for. 

I will look at your comments next up...

File: lib/gcs_signed_url/crypto.ex (hunk 1 above)
Comment (mruoss) [reply lib/gcs_signed_url/crypto.ex]: Okay I agree, the function was a bit overladed. I extracted the error formatter. But what speaks against pattern matching in the function declaration?

File: test/gcs_signed_url/crypto_test.exs (hunk 2 above)
Comment (mruoss) [reply test/gcs_signed_url/crypto_test.exs]: Partially agree. I want to make sure that the expected responses of the API return correct errors. But I can combine the several types of non-200 in a single test. Still, there's the case where the call fails completely that I would like to cover.

File: lib/gcs_signed_url/crypto.ex (hunk 1 above)
Comment (alexandrubagu) [reply lib/gcs_signed_url/crypto.ex]: I don't have any problem If you can make function head to fit in one line. I found very hard to read when you do something like this:
```elxiir
  def sign_via_api(
        string_to_sign,
        %Module1{
          some_param1: param1,
          some_param2: param2
        },
        %Module{
          some_param3: param3,
          some_param4: param4
        }
      ) do
      ...
  end
```

**Comment by mruoss** [comment]
Hey @alexandrubagu, any chance you can give this a quick review or shall I just merge? Thanks!

**Comment by alexandrubagu** [comment]
If you already use this code on production feel free to merge it. I'm not up to date regarding v4 URL signing

**Comment by mruoss** [comment]
This is not about v4 signing. It also works with v2. It's about where it is signed. Locally on the machine or remote (if the private key is not present on the machine). But okay, I will merge then.
